### ISSUE TITLE
Hierarchy type in typings doesn't have member levelName

### ISSUE THREAD

**Issue by ezk84** [flexmonster/ng-flexmonster#6]
Hierarchy in `flexmonster.d.ts` is defined as:

```ts
    interface Hierarchy {
        caption?: string;
        dimensionName?: string;
        filter?: Filter;
        sort?: string;
        sortOrder?: string[];
        uniqueName?: string;
        levels?: Level[];
    }
```

Is is missing a `levelName?: string;` in there to align with what's [documented here](https://www.flexmonster.com/api/slice-object/)

**Comment by mhalaida** [comment]
Hi @ezk84,

Thank you for posting your question.

The **levelName** property is not directly represented in the **Hierarchy** interface – instead, the hierarchy levels can be accessed through the corresponding **levels** property of the **Level** type:

```
interface Hierarchy {
    caption?: string;
    dimensionName?: string;
    filter?: Filter;
    sort?: string;
    sortOrder?: string[];
    uniqueName?: string;
    levels?: Level[];
}
```

This type in its turn provides a **uniqueName** property which represents the name of a particular level:

```
interface Level {
    caption: string;
    uniqueName: string;
}
```

The Level's **uniqueName** value equals the **levelName** value from the slice. 

Please let us know if this answers your question.

Best regards,
Mykhailo

**Comment by ezk84** [comment]
The problem I run into is when I am creating a Flexmonster.Report in TS to be passed into the Flexmonster pivot component, and I want to be type safe:

For example:
```ts
const report: Flexmonster.Report = {
    dataSource: {
        // ...
        mapping: { 
          // ... define hierarchy Geo with levels Type>State>Geo Name
        }
    },
    slice: {
        rows: [{
            uniqueName: 'Geo',
            levelName: 'Geo Name'
        }]
    }
};

// somewhere else 
pivot.setReport(report);
```

In this case TypeScript shows an error: 

> Type '{ uniqueName: string; levelName: string; }' is not assignable to type 'Hierarchy'.
>   Object literal may only specify known properties, and 'levelName' does not exist in type 'Hierarchy'.ts(2322)

I have tried setting instead the levels property:

```ts 
rows: [{
    uniqueName: 'Geo',
    levels: [{
        uniqueName: 'Geo Name',
    }],
}]
```
But then the report did not show just the level Geo Name, instead showing the whole Geo hierarchy.

**Comment by mhalaida** [comment]
Hi @ezk84,

Thank you for explaining your use case further and giving us some time to take a deeper look at this.

We've discussed this with our team and it seems like a mistake was made on our side when the ```Hierarchy``` interface was provided as a type for rows & columns of the Slice in our types definitions.

More specifically, this interface is only to be used for the ```getAllHierarchies()``` API call, which provides information about the hierarchies in a somewhat different structure from the one required in the Slice object.

With that in mind, we will change the ```flexmonster.d.ts``` definitions in a way that the separate ```SliceHierarchy``` interface is used to define fields in the Slice. This will allow setting the ```levelName``` with no exceptions, just like you've been trying to:

```
slice: {
    rows: [{
        uniqueName: 'Geo',
        levelName: 'Geo Name'
    }]
}
```
This fix will be available with our upcoming release **ETA Sep 21st**.

In the meantime, we would suggest using the ```any``` type for your Report object as a temporary workaround to avoid any exceptions.

Please let us know if this helps.

Best regards,
Mykhailo

**Comment by ezk84** [comment]
That makes sense, thank you!

**Comment by mhalaida** [comment]
Hi @ezk84,

We are happy to inform you that the `Slice` interface in `flexmonster.d.ts` type definitions was updated to provide the correct fields & measures structure.

This is available in the 2.8.16 version of Flexmonster: https://www.flexmonster.com/release-notes/.

You are welcome to update the component. Here is our updating to the latest version guide for assistance: https://www.flexmonster.com/doc/updating-to-the-latest-version/.
 
Please let us know if you have any other questions we can help you with.

Best regards,
Mykhailo
